src/System.IO.Compression/tests/DeflateStreamAsyncTests.cs

```diff
@@ -0,0 +1,166 @@
+    {
+        public static async Task<ManualSyncMemoryStream> GetStreamFromFileAsync(String testFile, bool sync = false, bool strip = false)
+        {
+            var baseStream = await StreamHelpers.CreateTempCopyStream(testFile);
+            if (strip)
+            {
+                baseStream = StripHeaderAndFooter.Strip(baseStream);
+            }
+
+            var ms = new ManualSyncMemoryStream(sync);
+            await baseStream.CopyToAsync(ms);
+
+            ms.Position = 0;
+            return ms;
+        }
+
+        public ManualResetEventSlim manualResetEvent = new ManualResetEventSlim(false);
+
+        private bool isSync = false;
+        public ManualSyncMemoryStream(bool sync) : base()
+        {
+            isSync = sync;
+        }
+
+        public void SetManualEvent()
+        {
+            manualResetEvent.Set();
+            // Logger.LogInformation("Signal");
+        }
+
+        public void ResetManualEvent()
+        {
+            manualResetEvent.Reset();
+        }
+
+        public override async Task<int> ReadAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (isSync)
+            {
+                manualResetEvent.Wait();
+            }
+            else
+            {
+                await Task.Run(() => manualResetEvent.Wait());
+            }
+ 
+            return await base.ReadAsync(array, offset, count, cancellationToken);
+        }
+
+        public override async Task WriteAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (isSync)
+            {
+                manualResetEvent.Wait(cancellationToken);
+            }
+            else
+            {
+                await Task.Run(() => manualResetEvent.Wait(cancellationToken));
+            }
+
+            await base.WriteAsync(array, offset, count, cancellationToken);
+        }
+    }
+
+    public partial class DeflateStreamTests
+    {
+        [Fact]
+        public static async Task OverlappingReadAsync()
+        {
+            byte[] buffer = new byte[32];
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "GZcompressed", "alice29.txt" + ".gz");
+            using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(testFilePath, false, true))
+            using (var unzip = new DeflateStream(readStream, CompressionMode.Decompress, true))
+            {
+                Task task = null;
+                try
+                {
+                    task = unzip.ReadAsync(buffer, 0, 32);
+                    Assert.Throws<InvalidOperationException>(() => { unzip.ReadAsync(buffer, 0, 32); }); // "overlapping read"
+                }
+                finally
+                {
+                   // clean up
+                    readStream.SetManualEvent();
+                    // internal assert that this should complete 
+                    Assert.True(task.Wait(100));
```

Comment:
This isn't really necessary anyways since what we care about is the exception above. Though I suppose testing that the first Async call should be able to complete is probably worth testing.